src/preview/topics.rs

```diff
@@ -0,0 +1,292 @@
+use momento_protos::cache_client::pubsub::{
+    self, pubsub_client::PubsubClient, PublishRequest, SubscriptionRequest,
+};
+use tonic::{codegen::InterceptedService, transport::Channel};
+
+use crate::MomentoError;
+use crate::{
+    endpoint_resolver::MomentoEndpointsResolver,
+    grpc::header_interceptor::HeaderInterceptor,
+    utils::{connect_channel_lazily, user_agent},
+    MomentoResult,
+};
+
+type ChannelType = InterceptedService<Channel, HeaderInterceptor>;
+
+pub struct Pubsub {
+    client: pubsub::pubsub_client::PubsubClient<ChannelType>,
+}
+
+/// Work with topics, publishing and subscribing.
+impl Pubsub {
+    pub fn connect(
+        auth_token: String,
+        momento_endpoint: Option<String>,
+        user_application_name: Option<&str>,
+    ) -> MomentoResult<Self> {
+        let momento_endpoints = MomentoEndpointsResolver::resolve(&auth_token, momento_endpoint)?;
+        let channel = connect_channel_lazily(&momento_endpoints.data_endpoint.url)?;
+        let authorized_channel = InterceptedService::new(
+            channel,
+            HeaderInterceptor::new(
+                &auth_token,
+                &user_agent(user_application_name.unwrap_or("sdk")),
+            ),
+        );
+        Ok(Self {
+            client: pubsub::pubsub_client::PubsubClient::new(authorized_channel),
+        })
+    }
+
+    /// Publish a value to a topic.
+    /// The cache is used as a namespace for your topics, and it needs to exist.
+    /// You don't create topics, you just start using them.
+    pub async fn publish(
+        &self,
+        cache_name: String,
+        topic: String,
+        value: impl IntoTopicValue,
+    ) -> Result<(), MomentoError> {
+        Pubsub::actually_publish(&mut self.client.clone(), cache_name, topic, value).await
+    }
+
+    /// Publish a value to a topic.
+    /// The cache is used as a namespace for your topics, and it needs to exist.
+    /// You don't create topics, you just start using them.
+    ///
+    /// Use this if you have &mut, as it will save you a small amount of overhead for reusing the client.
+    pub async fn publish_mut(
+        &mut self,
+        cache_name: String,
+        topic: String,
+        value: impl IntoTopicValue,
+    ) -> Result<(), MomentoError> {
```

Comment:
is it worth any treatment of what this error will look like if the connection is broken etc

Reply:
What do you have in mind? I'm reusing the same thing all our other apis use for error propagation (MomentoError) and it'll look the same as those.

Reply:
idk.  i just think this is a novel error case that i haven't seen with my own eyes yet, so i didn't know if there is anything specific that we should document about it.

Reply:
They all return Result<something, MomentoError> https://github.com/momentohq/client-sdk-rust/blob/main/src/simple_cache_client.rs#L418